src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertLinqQueryToForEachProvider<TQueryExpression> : CodeRefactoringProvider
+        where TQueryExpression : SyntaxNode
+    {
+        protected abstract string Title { get; }
+
+        protected abstract bool TryConvert(
+            TQueryExpression queryExpression,
+            SemanticModel semanticModel,
+            ISemanticFactsService semanticFacts,
+            CancellationToken cancellationToken,
+            out DocumentUpdateInfo documentUpdate);
+
+        protected abstract TQueryExpression FindNodeToRefactor(SyntaxNode root, TextSpan span);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var queryExpression = FindNodeToRefactor(root, context.Span);
+            if (queryExpression == null || queryExpression.ContainsDiagnostics)
```

Comment:
>ContainsDiagnostics [](start = 59, length = 19)

Does this cover semantic errors or are they handled elsewhere?

Reply:
We call _semanticModel.GetDiagnostics in the end (could take time) for the original query expression. It checks for semantic diagnostics.

---
In reply to: [176928116](https://github.com/dotnet/roslyn/pull/25362#discussion_r176928116) [](ancestors = 176928116)

Reply:
removed the check. agreed to make all the checks with GetDiagnostics filtered by errors only in the very end

---
In reply to: [178345599](https://github.com/dotnet/roslyn/pull/25362#discussion_r178345599) [](ancestors = 178345599,176928116)